patroni/quorum.py

```diff
@@ -0,0 +1,324 @@
+    increasing operation first. When decreasing either, the decreasing operation needs to be performed later.
+
+    Order of adding or removing nodes from sync and voters depends on the state of synchronous_standby_names:
+    When adding new nodes:
+        if sync (synchronous_standby_names) is empty:
+            add new nodes first to sync and than to voters when numsync_confirmed > 0
+        else:
+            add new nodes first to voters and than to sync
+    When removing nodes:
+        if sync (synchronous_standby_names) will become empty after removal:
+            first remove nodes from voters and than from sync
+        else:
+            first remove nodes from sync and than from voters. make voters empty if numsync_confirmed == 0"""
+
+    def __init__(self, leader: str, quorum: int, voters: Iterable[str],
+                 numsync: int, sync: Iterable[str], numsync_confirmed: int,
+                 active: Iterable[str], sync_wanted: int, leader_wanted: str) -> None:
+        self.leader = leader                          # The leader according to the `/sync` key
+        self.quorum = quorum                          # The number of nodes we need to check when doing leader race
+        self.voters = CaseInsensitiveSet(voters)      # Set of nodes we need to check (both stored in the /sync key)
+        self.numsync = numsync                        # The number of sync nodes in synchronous_standby_names
+        self.sync = CaseInsensitiveSet(sync)          # Set of nodes in synchronous_standby_names
+        # The number of nodes that are confirmed to reach safe LSN after adding them to `synchronous_standby_names`.
+        # We don't list them because it is known that they are always included into active.
+        self.numsync_confirmed = numsync_confirmed
+        self.active = CaseInsensitiveSet(active)      # Set of active nodes from `pg_stat_replication`
+        self.sync_wanted = sync_wanted                # The desired number of sync nodes
+        self.leader_wanted = leader_wanted or leader  # The desired leader
+
+    def check_invariants(self) -> None:
+        """Checks invatiant of synchronous_standby_names and /sync key in DCS.
+
+        We need to verify that subset of nodes that can acknowledge a commit overlaps
+        with any subset of nodes that can achieve quorum to promote a new leader.
+
+        :raises QuorumError: in case of broken state"""
+
+        voters = self.voters | CaseInsensitiveSet([self.leader])
+        sync = self.sync | CaseInsensitiveSet([self.leader_wanted])
+        if self.voters and not (len(voters | sync) <= self.quorum + self.numsync + 1):
+            raise QuorumError("Quorum and sync not guaranteed to overlap: nodes %d >= quorum %d + sync %d" %
+                              (len(voters | sync), self.quorum, self.numsync))
+        if not (voters <= sync or sync <= voters):
```

Comment:
we need this condition to be described in the docstring
\+ it is maybe more clear to use `.issubset()` here